src/BotPlutusInterface/Balance.hs

```diff
@@ -326,39 +398,30 @@ handleNonAdaChange changeAddr utxos tx =
         then Right $ if Value.isZero nonAdaChange then tx else tx {txOutputs = outputs}
         else Left "Not enough inputs to balance tokens."
 
-hasChangeUTxO :: Address -> Tx -> Bool
-hasChangeUTxO changeAddr tx =
-  any ((==) changeAddr . Tx.txOutAddress) $ txOutputs tx
-
--- | Adds ada change to a transaction, assuming there is already an output going to ownPkh. Otherwise, this is identity
-addAdaChange :: Address -> Integer -> Tx -> Tx
-addAdaChange _ 0 tx = tx
-addAdaChange changeAddr change tx =
-  tx
-    { txOutputs =
-        modifyFirst
-          ((==) changeAddr . Tx.txOutAddress)
-          (fmap $ addValueToTxOut $ Ada.lovelaceValueOf change)
-          (txOutputs tx)
-    }
-
-consJust :: forall (a :: Type). Maybe a -> [a] -> [a]
-consJust (Just x) = (x :)
-consJust _ = id
-
-{- | Modifies the first element matching a predicate, or, if none found, call the modifier with Nothing
- Calling this function ensures the modifier will always be run once
+{- | `addAdaChange` checks if `bcSeparateChange` is true,
+      if it is then we add the ada change to seperate `TxOut` at changeAddr that contains only ada,
+      else we add it to any `TxOut` present at changeAddr.
 -}
-modifyFirst ::
-  forall (a :: Type).
-  -- | Predicate for value to update
-  (a -> Bool) ->
-  -- | Modifier, input Maybe representing existing value (or Nothing if missing), output value representing new value (or Nothing to remove)
-  (Maybe a -> Maybe a) ->
-  [a] ->
-  [a]
-modifyFirst _ m [] = m Nothing `consJust` []
-modifyFirst p m (x : xs) = if p x then m (Just x) `consJust` xs else x : modifyFirst p m xs
+addAdaChange :: BalanceConfig -> Address -> Integer -> Tx -> Tx
+addAdaChange _ _ 0 tx = tx
+addAdaChange balanceCfg changeAddr change tx
+  | bcSeparateChange balanceCfg =
+    tx
+      { txOutputs =
+          List.reverse $
```

Comment:
Why is the list being reversed twice?

Reply:
Actually, here's how the logic work:

If the flag `bcSeparateChange` is `True` then we create a separate change utxo at the changeAddr, this is done by `addOutput`, this adds *ada only* utxo at the *end* of the list, so this is the change utxo where all the ada change should go. There might be some other utxos present in the TxOutput at the changeAddr previously, we don't want to touch those utxos. Hence, we reverse the list and check for the "first" *ada* only utxo.

Reply:
I will make sure to add the docs for these once everything is finalized.

Reply:
This is with separate change, if we don't have seperate change, it'd add to `out1` (plutus will build [out2, out1]).
I assume separate change is the expected behaviour for normal balancing though, right?

Reply:
So, here's how it works if the `bcSeparateChange` flag is `False`. Instead of searching for *last ada only* utxo, we just search for *any* utxo present at the changeAddr.

Here's how the everything will work if `bcSeparateChange` is `False`:
1. check if the there's already an Tx output present at the changeAddr if not create one, in this case we don't create because `out1` is already present.
2. search of *first* utxo present at the changeAddr and add the change to it. Again, this doesn't depend on how plutus builds the Tx.

And the default value of `bcSeparateChange` is False. Because that's how it was previously.